aarch64: brgemm: add jit impl. for sve_512

Koji-Kurihara opened the pull request:
# Description

This patch adds jit implementation of brgemm for AArch64 SVE512.

Please merge after applying the following pull requests.
#1461
#1462 

# Checklist

## General

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [x] Have you formatted the code using clang-format?

mgouicem reviewed (commented):
I think you forgot to add the inner_product implementation to cpu_inner_product_list.cpp.

Also could you clarify why you are keeping the whole AMX logic from x64? One of the goodness of the brgemm approach is that it should allow you to "just" reimplement the [brgemm API ](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/x64/brgemm/brgemm.hpp) in the best way possible for your arch, and then reuse the same driver across architecture. I think this would simplify the whole brgemm support on the ARM side.

mgouicem commented on src/cpu/aarch64/jit_brgemm_conv.cpp:
```diff
@@ -0,0 +1,1491 @@
+                    64)
+                != 0) {
+            //amx_tile_configure(brg_kernel_palettes_[brg_idx].a);
+            printf("%s, %d, unimplemented\n", __FILE__, __LINE__);
```
we typically don't do printf inside code.

Koji-Kurihara replied:
I'm sorry about this.
I removed the printf.

densamoilov commented on src/cpu/aarch64/amx_tile_configure.hpp:
```diff
@@ -0,0 +1,37 @@
+
+static constexpr size_t AMX_PALETTE_SIZE = 64;
+status_t DNNL_API amx_tile_configure(const char palette[AMX_PALETTE_SIZE]);
+status_t DNNL_API amx_tile_release();
```
Do you use these or it's just for simplicity?

Koji-Kurihara replied:
On the ARM side, amx_tile_configure.hpp is not used, so I removed it.

Koji-Kurihara commented:
Thank you for your comment.
> I think you forgot to add the inner_product implementation to cpu_inner_product_list.cpp.

I have added the inner_product implementation to cpu_inner_product_list.cpp.

> Also could you clarify why you are keeping the whole AMX logic from x64? One of the goodness of the brgemm approach is that it should allow you to "just" reimplement the [brgemm API ](https://github.com/oneapi-> src/oneDNN/blob/master/src/cpu/x64/brgemm/brgemm.hpp)in the best way possible for your arch, and then reuse the same driver across architecture. I think this would simplify the whole brgemm support on the ARM side.

I temporarily left the AMX implementation to eliminate build errors. 
Since the ARM side does not use the AMX implementation, 
I removed the AMX implementation.

igorsafo commented:
Hi @Koji-Kurihara ,

There were recent changes to quantization support described in this RFC: [20220201-quantization-scaling](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20220201-quantization-scaling )

Please review the x86 version of brgemm and update aarch64 version to make it work for int8.

The main changes are:
- Now scaling relies on per argument scales instead of a single output scale
   - To reuse brgemm in other primitives currently it still takes a single pointer to `scales`, but a primitive that uses brgemm should precompute that scale by multiplying src_scale and wei_scale.
   - dst scale is not yet supported for brgemm kernels
- Bias is aplied after dequantization (after appling input scales) and doesn't need to be quantized by user
- Scaling is runtime only, so scale factors are passed at execution time as separate memory objects and they are not known at creation time.

igorsafo commented:
Hi @Koji-Kurihara @kawakami-k ,
Do you plan to make any additional changes to this PR?

Koji-Kurihara commented:
Hi @igorsafo ,
I'm sorry.
I have not been able to work on this issue due to lack of time.
We will work on this issue before the Branch cutoff in v3.1.

igorsafo commented:
@Koji-Kurihara Thanks for the update!

Koji-Kurihara commented:
Hi @igorsafo, 
I am developing brgemm to work with int8.
However, it will take more time to complete.
So, could you please merge the current source code implementing brgemm that works with f32?
I will send out another pull request for brgemm that works with int8 as soon as I finish development.

mgouicem commented:
Maybe I am missing something, but I don't see the changes mentioned in your comments (removing amx_tileconfigure headers, printf, ...). Did you push those changes?

Koji-Kurihara commented:
Hi @mgouicem, 
Can you see [bcd4f7d](https://github.com/oneapi-src/oneDNN/pull/1463/commits/bcd4f7d7f605a537c47d376faa827b46583581d1)?
This commit ID contains those changes.

mgouicem reviewed (approved):
Thanks I now see the changes.
Also, could you please update the copyright header to use the proper year (2023)? Thanks.

mgouicem commented on src/cpu/cpu_inner_product_list.cpp:
```diff
@@ -27,6 +27,10 @@
+#endif
 #if DNNL_AARCH64 && DNNL_AARCH64_USE_ACL
 #include "cpu/aarch64/acl_inner_product.hpp"
 using namespace dnnl::impl::cpu::aarch64;
```
(nitpicking) would  it make sense to remove the now redundant `using namespace dnnl::impl::cpu::aarch64;` that is under USE_ACL ?

Koji-Kurihara commented:
Hi @mgouicem,
Thank you for your comment.
I updated the copyright header and removed redundant descriptions.
[a796e94](https://github.com/oneapi-src/oneDNN/pull/1463/commits/a796e944e0c00f62d910fad770df161d7049a0d6) contains these changes.

mgouicem commented:
One last thing: please fix formatting using [clang-format](https://dev.azure.com/mkldnn/dnnl/_build/results?buildId=1550&view=logs&j=0bf7d65c-1444-5302-7319-c6947569d8d5&t=85a4fd38-810c-57ff-5840-0f66a7da7b77).

Koji-Kurihara commented:
Thank you for your comment.
I fixed the formatting using clang-format in [f4e74db](https://github.com/oneapi-src/oneDNN/pull/1463/commits/f4e74dbede986bdc3676bda123598a0baecc2f23).

kawakami-k commented:
I'm modifying  SVE512 implementation to match the latest x64 implementation. Please wait a week.

vineelabhinav commented:
Hello @igorsafo @mgouicem @densamoilov,
I have raised new PR for the same brgemm jit implementation support.
Link to PR: https://github.com/oneapi-src/oneDNN/pull/1818
Requesting you to please close this PR.
Thanks

igorsafo commented:
Link to the new version: https://github.com/oneapi-src/oneDNN/pull/1818
